File: src/main/java/com/mpe85/grampa/rule/impl/RegexRule.java
package com.mpe85.grampa.rule.impl;

import java.util.Objects;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import com.google.common.base.MoreObjects.ToStringHelper;
import com.mpe85.grampa.rule.RuleContext;

/**
 * A regular expression rule implementation.
 * 
 * @author mpe85
 *
 * @param <T>
 *            the type of the stack elements
 */
public class RegexRule<T> extends AbstractRule<T> {
	
	private final Pattern pattern;
	
	/**
	 * C'tor. Constructs a regex rule using an already compiled {@link Pattern}.
	 * 
	 * @param pattern
	 *            a compiled regular expression
	 */
	public RegexRule(final Pattern pattern) {
		this.pattern = pattern;
	}
	
	/**
	 * C'tor. Constructs a regex rule using a regex string.
	 * 
	 * @param regex
	 *            a string containing a regular expression
	 */
	public RegexRule(final String regex) {
		pattern = Pattern.compile(regex);
	}
	
	@Override
	public boolean match(final RuleContext<T> context) {
		final Matcher regexMatcher = pattern.matcher(context.getRestOfInput());
		return regexMatcher.lookingAt() && context.advanceIndex(regexMatcher.end());
	}
	
	@Override
	public int hashCode() {
		return Objects.hash(super.hashCode(), pattern);
	}
	
	@Override
	public boolean equals(final Object obj) {
		if (obj != null && getClass() == obj.getClass()) {
			final RegexRule<?> other = (RegexRule<?>) obj;
			return super.equals(other)
					&& Objects.equals(pattern, other.pattern);
		}
		return false;
	}
	
	@Override
	protected ToStringHelper toStringHelper() {
		return super.toStringHelper()
				.add("pattern", pattern);
	}
	
}

File: src/main/kotlin/com/mpe85/grampa/rule/impl/RegexRule.kt
package com.mpe85.grampa.rule.impl

import com.mpe85.grampa.context.ParserContext
import com.mpe85.grampa.util.checkEquality
import com.mpe85.grampa.util.stringify
import java.util.Objects.hash
import java.util.regex.Pattern
import java.util.regex.Pattern.compile

/**
 * A regular expression rule implementation.
 *
 * @author mpe85
 * @param[T] The type of the stack elements
 * @property[pattern] A compiled regular expression
 */
public class RegexRule<T>(private val pattern: Pattern) : AbstractRule<T>() {

    /**
     * Construct a regex rule using a regex string.
     *
     * @param[regex] A string containing a regular expression
     */
    public constructor(regex: String) : this(compile(regex))

    override fun match(context: ParserContext<T>): Boolean = pattern.matcher(context.restOfInput).let { matcher ->
        matcher.lookingAt() && context.advanceIndex(matcher.end())
    }

    override fun hashCode(): Int = hash(super.hashCode(), pattern)
    override fun equals(other: Any?): Boolean = checkEquality(other, { super.equals(other) }, { it.pattern })
    override fun toString(): String = stringify("pattern" to pattern)
}
